# Table of Contents
### 1. Import libraries and data
### 2. Data Cleaning

## 1. Import libraries and data

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Create path
path = r'/Users/ianfleming/Desktop/World Happiness Report Analysis'

In [3]:
# Import Data
df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'World Happiness Report 2005 - 2023.csv'),index_col = False)

In [4]:
# Check import
df.head()

Country Name Regional Indicator  Year  Life Ladder  Log GDP Per Capita  \
0  Afghanistan         South Asia  2008     3.723590            7.350416   
1  Afghanistan         South Asia  2009     4.401778            7.508646   
2  Afghanistan         South Asia  2010     4.758381            7.613900   
3  Afghanistan         South Asia  2011     3.831719            7.581259   
4  Afghanistan         South Asia  2012     3.782938            7.660506   

   Social Support  Healthy Life Expectancy At Birth  \
0        0.450662                         50.500000   
1        0.552308                         50.799999   
2        0.539075                         51.099998   
3        0.521104                         51.400002   
4        0.520637                         51.700001   

   Freedom To Make Life Choices  Generosity  Perceptions Of Corruption  \
0                      0.718114    0.167652                   0.881686   
1                      0.678896    0.190809                   0.850035   
2                      0.600127    0.121316                   0.706766   
3                      0.495901    0.163571                   0.731109   
4                      0.530935    0.237588                   0.775620   

   Positive Affect  Negative Affect  Confidence In National Government  
0         0.414297         0.258195                           0.612072  
1         0.481421         0.237092                           0.611545  
2         0.516907         0.275324                           0.299357  
3         0.479835         0.267175                           0.307386  
4         0.613513         0.267919                           0.435440

In [5]:
df.tail()

Country Name  Regional Indicator  Year  Life Ladder  Log GDP Per Capita  \
2194     Zimbabwe  Sub-Saharan Africa  2018     3.616480            7.783066   
2195     Zimbabwe  Sub-Saharan Africa  2019     2.693523            7.697755   
2196     Zimbabwe  Sub-Saharan Africa  2020     3.159802            7.596050   
2197     Zimbabwe  Sub-Saharan Africa  2021     3.154578            7.656878   
2198     Zimbabwe  Sub-Saharan Africa  2022     3.296220            7.670123   

      Social Support  Healthy Life Expectancy At Birth  \
2194        0.775388                         52.625000   
2195        0.759162                         53.099998   
2196        0.717243                         53.575001   
2197        0.685151                         54.049999   
2198        0.666172                         54.525002   

      Freedom To Make Life Choices  Generosity  Perceptions Of Corruption  \
2194                      0.762675   -0.051219                   0.844209   
2195                      0.631908   -0.047464                   0.830652   
2196                      0.643303    0.006313                   0.788523   
2197                      0.667636   -0.075575                   0.756945   
2198                      0.651987   -0.069513                   0.752632   

      Positive Affect  Negative Affect  Confidence In National Government  
2194         0.657524         0.211726                           0.550508  
2195         0.658434         0.235354                           0.456455  
2196         0.660658         0.345736                           0.577302  
2197         0.609917         0.241682                           0.665703  
2198         0.640609         0.191350                                NaN

In [6]:
# Check shape
df.shape

(2199, 13)

In [7]:
# Check info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2199 entries, 0 to 2198
Data columns (total 13 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Country Name                       2199 non-null   object 
 1   Regional Indicator                 2087 non-null   object 
 2   Year                               2199 non-null   int64  
 3   Life Ladder                        2199 non-null   float64
 4   Log GDP Per Capita                 2179 non-null   float64
 5   Social Support                     2186 non-null   float64
 6   Healthy Life Expectancy At Birth   2145 non-null   float64
 7   Freedom To Make Life Choices       2166 non-null   float64
 8   Generosity                         2126 non-null   float64
 9   Perceptions Of Corruption          2083 non-null   float64
 10  Positive Affect                    2175 non-null   float64
 11  Negative Affect                    2183 non-null   float

In [8]:
# Standardise column names
df.columns = (df.columns
    .str.strip()                              # remove trailing spaces
    .str.lower()                              # lowercase everything
    .str.replace(" ", "_")                    # replace spaces with underscores
    .str.replace(r"[^\w\s]", "", regex=True))  # remove special characters


In [9]:
# Check changes
df.head()

country_name regional_indicator  year  life_ladder  log_gdp_per_capita  \
0  Afghanistan         South Asia  2008     3.723590            7.350416   
1  Afghanistan         South Asia  2009     4.401778            7.508646   
2  Afghanistan         South Asia  2010     4.758381            7.613900   
3  Afghanistan         South Asia  2011     3.831719            7.581259   
4  Afghanistan         South Asia  2012     3.782938            7.660506   

   social_support  healthy_life_expectancy_at_birth  \
0        0.450662                         50.500000   
1        0.552308                         50.799999   
2        0.539075                         51.099998   
3        0.521104                         51.400002   
4        0.520637                         51.700001   

   freedom_to_make_life_choices  generosity  perceptions_of_corruption  \
0                      0.718114    0.167652                   0.881686   
1                      0.678896    0.190809                   0.850035   
2                      0.600127    0.121316                   0.706766   
3                      0.495901    0.163571                   0.731109   
4                      0.530935    0.237588                   0.775620   

   positive_affect  negative_affect  confidence_in_national_government  
0         0.414297         0.258195                           0.612072  
1         0.481421         0.237092                           0.611545  
2         0.516907         0.275324                           0.299357  
3         0.479835         0.267175                           0.307386  
4         0.613513         0.267919                           0.435440

Checking For Missing Values

In [10]:
# Check info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2199 entries, 0 to 2198
Data columns (total 13 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   country_name                       2199 non-null   object 
 1   regional_indicator                 2087 non-null   object 
 2   year                               2199 non-null   int64  
 3   life_ladder                        2199 non-null   float64
 4   log_gdp_per_capita                 2179 non-null   float64
 5   social_support                     2186 non-null   float64
 6   healthy_life_expectancy_at_birth   2145 non-null   float64
 7   freedom_to_make_life_choices       2166 non-null   float64
 8   generosity                         2126 non-null   float64
 9   perceptions_of_corruption          2083 non-null   float64
 10  positive_affect                    2175 non-null   float64
 11  negative_affect                    2183 non-null   float

There is a lot of missing values

## 2. Data Cleaning

In [11]:
# Check missing values
df.isnull().sum()

country_name                           0
regional_indicator                   112
year                                   0
life_ladder                            0
log_gdp_per_capita                    20
social_support                        13
healthy_life_expectancy_at_birth      54
freedom_to_make_life_choices          33
generosity                            73
perceptions_of_corruption            116
positive_affect                       24
negative_affect                       16
confidence_in_national_government    361
dtype: int64

ten of the 13 variables have missing values that need addressing.

Addressing Missing Values

There is a large amount of missing values in confidence_in_national_government column. I don't want to replace the missing values so creating a flag column.

FIRST MISSING

In [12]:
# Create missing flag column
df["conf_gov_missing"] = df["confidence_in_national_government"].isnull().astype(int)


In [13]:
# Check flag column
df[["confidence_in_national_government", "conf_gov_missing"]].head(10)

confidence_in_national_government  conf_gov_missing
0                           0.612072                 0
1                           0.611545                 0
2                           0.299357                 0
3                           0.307386                 0
4                           0.435440                 0
5                           0.482847                 0
6                           0.409048                 0
7                           0.260557                 0
8                           0.324990                 0
9                           0.261179                 0

Addressing regional_indicator column which has 112 missing values

SECOND MISSING

In [14]:
# Fill missing values with "Unknown"
df["regional_indicator"] = df["regional_indicator"].fillna("Unknown")

In [15]:
# Check for missing values
df["regional_indicator"].isnull().sum()

0

gdp per capita has 20 missing values. As it is not so many and is numeric, I will impute the mean

THIRD MISSING

In [16]:
# Impute GDP with the mean
df["log_gdp_per_capita"] = df["log_gdp_per_capita"].fillna(df["log_gdp_per_capita"].mean())

In [17]:
# Check values
df["log_gdp_per_capita"].isnull().sum()

0

FOURTH MISSING

In [18]:
df[df["healthy_life_expectancy_at_birth"].isnull()]

country_name          regional_indicator  year  \
793   Hong Kong S.A.R. of China                   East Asia  2006   
794   Hong Kong S.A.R. of China                   East Asia  2008   
795   Hong Kong S.A.R. of China                   East Asia  2009   
796   Hong Kong S.A.R. of China                   East Asia  2010   
797   Hong Kong S.A.R. of China                   East Asia  2011   
798   Hong Kong S.A.R. of China                   East Asia  2012   
799   Hong Kong S.A.R. of China                   East Asia  2014   
800   Hong Kong S.A.R. of China                   East Asia  2016   
801   Hong Kong S.A.R. of China                   East Asia  2017   
802   Hong Kong S.A.R. of China                   East Asia  2019   
803   Hong Kong S.A.R. of China                   East Asia  2020   
804   Hong Kong S.A.R. of China                   East Asia  2021   
1031                     Kosovo  Central and Eastern Europe  2007   
1032                     Kosovo  Central and Eastern Europe  2008   
1033                     Kosovo  Central and Eastern Europe  2009   
1034                     Kosovo  Central and Eastern Europe  2010   
1035                     Kosovo  Central and Eastern Europe  2011   
1036                     Kosovo  Central and Eastern Europe  2012   
1037                     Kosovo  Central and Eastern Europe  2013   
1038                     Kosovo  Central and Eastern Europe  2014   
1039                     Kosovo  Central and Eastern Europe  2015   
1040                     Kosovo  Central and Eastern Europe  2016   
1041                     Kosovo  Central and Eastern Europe  2017   
1042                     Kosovo  Central and Eastern Europe  2018   
1043                     Kosovo  Central and Eastern Europe  2019   
1044                     Kosovo  Central and Eastern Europe  2020   
1045                     Kosovo  Central and Eastern Europe  2021   
1046                     Kosovo  Central and Eastern Europe  2022   
1755          Somaliland region                     Unknown  2009   
1756          Somaliland region                     Unknown  2010   
1757          Somaliland region                     Unknown  2011   
1758          Somaliland region                     Unknown  2012   
1833         State of Palestine                     Unknown  2011   
1834         State of Palestine                     Unknown  2012   
1835         State of Palestine                     Unknown  2013   
1836         State of Palestine                     Unknown  2014   
1837         State of Palestine                     Unknown  2015   
1838         State of Palestine                     Unknown  2016   
1839         State of Palestine                     Unknown  2017   
1840         State of Palestine                     Unknown  2018   
1841         State of Palestine                     Unknown  2019   
1842         State of Palestine                     Unknown  2022   
1888   Taiwan Province of China                   East Asia  2011   
1889   Taiwan Province of China                   East Asia  2012   
1890   Taiwan Province of China                   East Asia  2013   
1891   Taiwan Province of China                   East Asia  2014   
1892   Taiwan Province of China                   East Asia  2015   
1893   Taiwan Province of China                   East Asia  2016   
1894   Taiwan Province of China                   East Asia  2017   
1895   Taiwan Province of China                   East Asia  2018   
1896   Taiwan Province of China                   East Asia  2019   
1897   Taiwan Province of China                   East Asia  2020   
1898   Taiwan Province of China                   East Asia  2021   
1899   Taiwan Province of China                   East Asia  2022   

      life_ladder  log_gdp_per_capita  social_support  \
793      5.511187           10.746400        0.812178   
794      5.137262           10.815522        0.840222   
795      5.397056           10.788470        0.83471

healthy_life_expectancy_at_birth has 54 NaNs so I will impute them with the mean

In [19]:
# Impute life expectancy with the mean
df["healthy_life_expectancy_at_birth"] = df["healthy_life_expectancy_at_birth"].fillna(df["healthy_life_expectancy_at_birth"].mean())

In [20]:
# Check impute
df["healthy_life_expectancy_at_birth"].isnull().sum()

0

FIFTH MISSING

In [21]:
# Check missing values for social_support
df[df["social_support"].isnull()]

country_name            regional_indicator  year  life_ladder  \
29                 Algeria  Middle East and North Africa  2010     5.463567   
126                Bahrain  Middle East and North Africa  2014     6.165134   
328                 Canada         North America and ANZ  2007     7.481753   
524               Djibouti                       Unknown  2010     5.005811   
1053                Kuwait  Middle East and North Africa  2014     6.180139   
1323               Morocco  Middle East and North Africa  2010     4.383247   
1485                  Oman                       Unknown  2011     6.852982   
1599                 Qatar                       Unknown  2010     6.849653   
1602                 Qatar                       Unknown  2015     6.374529   
1966               Tunisia  Middle East and North Africa  2009     5.025470   
2045  United Arab Emirates  Middle East and North Africa  2014     6.539855   
2072         United States         North America and ANZ  2007     7.512688   
2149               Vietnam                Southeast Asia  2017     5.175279   

      log_gdp_per_capita  social_support  healthy_life_expectancy_at_birth  \
29              9.306355             NaN                         65.500000   
126            10.801981             NaN                         65.959999   
328            10.733970             NaN                         70.620003   
524             7.934187             NaN                         54.599998   
1053           10.925643             NaN                         70.000000   
1323            8.672808             NaN                         62.500000   
1485           10.538620             NaN                         62.340000   
1599           11.551208             NaN                         64.699997   
1602           11.532454             NaN                         66.400002   
1966            9.237887             NaN                         66.220001   
2045           11.071845             NaN                         65.339996   
2072           10.931180             NaN                         66.760002   
2149            9.110596             NaN                         65.099998   

      freedom_to_make_life_choices  generosity  perceptions_of_corruption  \
29                        0.592696   -0.209753                   0.618038   
126                            NaN         NaN                        NaN   
328                       0.930341    0.245596                   0.405608   
524                       0.763730   -0.068953                   0.596910   
1053                           NaN         NaN                        NaN   
1323                      0.662900   -0.159650                   0.900453   
1485                      0.916293    0.009673                        NaN   
1599                           NaN    0.096500                        NaN   
1602                           NaN         NaN                        NaN   
1966                      0.781496   -0.124444                   0.722211   
2045                           NaN         NaN                        NaN   
2072                      0.871904    0.192796                   0.633035   
2149                           NaN         NaN                        NaN   

      positive_affect  negative_affect  confidence_in_national_government  \
29                NaN              NaN                                NaN   
126               NaN              NaN                                NaN   
328          0.811577         0.256810                           0.637824   
524               NaN              NaN                           0.692278   
1053              NaN              NaN                                NaN   
1323              NaN              NaN                                NaN   
1485              NaN         0.295164                                NaN   
1599              NaN              NaN                                NaN   
1602              NaN              NaN    

social_support has 13 missing values so will impute the mean

In [22]:
# Impute social support with the mean
df["social_support"] = df["social_support"].fillna(df["social_support"].mean())


In [23]:
# Check impute
df["social_support"].isnull().sum()

0

SIXTH MISSING

In [24]:
# Check missing values for freedom_to_make_life_choices
df[df["freedom_to_make_life_choices"].isnull()]

country_name                  regional_indicator  year  \
32                 Algeria        Middle East and North Africa  2014   
33                 Algeria        Middle East and North Africa  2016   
126                Bahrain        Middle East and North Africa  2014   
293               Cambodia                      Southeast Asia  2006   
381                  China                           East Asia  2006   
382                  China                           East Asia  2007   
389                  China                           East Asia  2014   
390                  China                           East Asia  2015   
391                  China                           East Asia  2016   
562                  Egypt        Middle East and North Africa  2008   
885                   Iraq        Middle East and North Africa  2013   
980                 Jordan        Middle East and North Africa  2005   
982                 Jordan        Middle East and North Africa  2008   
1032                Kosovo          Central and Eastern Europe  2008   
1053                Kuwait        Middle East and North Africa  2014   
1079                  Laos                      Southeast Asia  2012   
1167            Madagascar                  Sub-Saharan Africa  2006   
1599                 Qatar                             Unknown  2010   
1602                 Qatar                             Unknown  2015   
1648          Saudi Arabia        Middle East and North Africa  2005   
1664          Saudi Arabia        Middle East and North Africa  2022   
1912            Tajikistan  Commonwealth of Independent States  2018   
1913            Tajikistan  Commonwealth of Independent States  2019   
1914            Tajikistan  Commonwealth of Independent States  2020   
1915            Tajikistan  Commonwealth of Independent States  2021   
1996          Turkmenistan  Commonwealth of Independent States  2009   
1997          Turkmenistan  Commonwealth of Independent States  2011   
2045  United Arab Emirates        Middle East and North Africa  2014   
2107            Uzbekistan  Commonwealth of Independent States  2009   
2108            Uzbekistan  Commonwealth of Independent States  2010   
2146               Vietnam                      Southeast Asia  2014   
2147               Vietnam                      Southeast Asia  2015   
2149               Vietnam                      Southeast Asia  2017   

      life_ladder  log_gdp_per_capita  social_support  \
32       6.354898            9.355415        0.818189   
33       5.340854            9.383312        0.748588   
126      6.165134           10.801981        0.810679   
293      3.568745            7.746443        0.793081   
381      4.560495            8.696139        0.747011   
382      4.862862            8.823968        0.810852   
389      5.195619            9.380201        0.820366   
390      5.303878            9.442431        0.793734   
391      5.324956            9.502946        0.741703   
562      4.631741            9.134907        0.738364   
885      4.725017            9.159455        0.728285   
980      6.294660            9.282825        0.920013   
982      4.930058            9.353609        0.766224   
1032     5.521660            8.858291        0.883843   
1053     6.180139           10.925643        0.810679   
1079     4.876085            8.600584        0.692628   
1167     3.979751            7.351137        0.711135   
1599     6.849653           11.551208        0.810679   
1602     6.374529           11.532454        0.810679   
1648     7.079644           10.674833        0.867819   
1664     6.381610           10.780560        0.900105   
1912     5.497469            8.133027        0.875243   
1913     5.464015            8.181799        0.879823   
1914     5.373399            8.203014        0.789745   
1915     5.286824            8.269583        0.882889   
1996     6.567713            8.955209        0.923846   
1997     5.791755         

In [25]:
# Impute freedom with the mean
df["freedom_to_make_life_choices"] = df["freedom_to_make_life_choices"].fillna(
    df["freedom_to_make_life_choices"].mean())


In [26]:
# Check impute
df["freedom_to_make_life_choices"].isnull().sum()

0

SEVENTH MISSING

Generosity has 73 missing values

In [27]:
# Impute generosity with the mean
df["generosity"] = df["generosity"].fillna(df["generosity"].mean())

In [28]:
# Check
df["generosity"].isnull().sum()

0

EIGTH MISSING

Perceptions_of_corruption has 116 misisng values

In [29]:
# Impute perceptions of corruption with the mean
df["perceptions_of_corruption"] = df["perceptions_of_corruption"].fillna(
    df["perceptions_of_corruption"].mean())

In [30]:
# Check
df["perceptions_of_corruption"].isnull().sum()

0

NINTH MISSING

Positive_affect has 19 missing values

In [31]:
# Impute positive_affect with the mean
df["positive_affect"] = df["positive_affect"].fillna(df["positive_affect"].mean())

In [32]:
# Check impute
df["positive_affect"].isnull().sum()

0

TENTH MISSING

Negative_affect has 16 missing

In [33]:
# Impute negative affect with the mean
df["negative_affect"] = df["negative_affect"].fillna(df["negative_affect"].mean())

In [34]:
# Check impute
df["negative_affect"].isnull().sum()

0

Now the data has 0 missing values except in the column "confidence_in_national_government" which I have created a flag column called "conf_gov_missing"

In [35]:
# Check missing values
df.isnull().sum()

country_name                           0
regional_indicator                     0
year                                   0
life_ladder                            0
log_gdp_per_capita                     0
social_support                         0
healthy_life_expectancy_at_birth       0
freedom_to_make_life_choices           0
generosity                             0
perceptions_of_corruption              0
positive_affect                        0
negative_affect                        0
confidence_in_national_government    361
conf_gov_missing                       0
dtype: int64

CHECK FOR DUPLICATES

In [36]:
# Check duplicate rows
df.duplicated().sum()

0

DESCRIPTIVE STATISTICS

In [37]:
# Descriptive stats for all numeric columns
df.describe().T

count         mean       std          min  \
year                               2199.0  2014.161437  4.718736  2005.000000   
life_ladder                        2199.0     5.479226  1.125529     1.281271   
log_gdp_per_capita                 2199.0     9.389766  1.148128     5.526723   
social_support                     2199.0     0.810679  0.120594     0.228217   
healthy_life_expectancy_at_birth   2199.0    63.294583  6.815805     6.720000   
freedom_to_make_life_choices       2199.0     0.747858  0.139094     0.257534   
generosity                         2199.0     0.000096  0.158385    -0.337527   
perceptions_of_corruption          2199.0     0.745195  0.180867     0.035198   
positive_affect                    2199.0     0.652143  0.105342     0.178886   
negative_affect                    2199.0     0.271501  0.086559     0.082737   
confidence_in_national_government  1838.0     0.483999  0.193071     0.068769   
conf_gov_missing                   2199.0     0.164166  0.370510     0.000000   

                                           25%          50%          75%  \
year                               2010.000000  2014.000000  2018.000000   
life_ladder                           4.646750     5.432437     6.309460   
log_gdp_per_capita                    8.505131     9.491772    10.366237   
social_support                        0.747512     0.834280     0.904607   
healthy_life_expectancy_at_birth     59.412500    64.879997    68.329998   
freedom_to_make_life_choices          0.658615     0.766945     0.857775   
generosity                           -0.107358    -0.016512     0.086938   
perceptions_of_corruption             0.697610     0.791269     0.865565   
positive_affect                       0.572742     0.661645     0.736762   
negative_affect                       0.208157     0.261508     0.322382   
confidence_in_national_government     0.332549     0.467140     0.618846   
conf_gov_missing                      0.000000     0.000000     0.000000   

                                           max  
year                               2022.000000  
life_ladder                           8.018934  
log_gdp_per_capita                   11.663788  
social_support                        0.987343  
healthy_life_expectancy_at_birth     74.474998  
freedom_to_make_life_choices          0.985178  
generosity                            0.702708  
perceptions_of_corruption             0.983276  
positive_affect                       0.883586  
negative_affect                       0.704590  
confidence_in_national_government     0.993604  
conf_gov_missing                      1.000000

In [38]:
# Check shape of data
df.shape

(2199, 14)

In [39]:
# Export cleaned data
df.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'world_happiness_report.csv'))